x509/src/attr.rs

```diff
@@ -87,3 +91,164 @@ pub struct AttributeTypeAndValue<'a> {
+        match (self.state, c) {
+            (Escape::Hex(p), b'0'..=b'9') => self.push(p | (c - b'0')),
+            (Escape::Hex(p), b'a'..=b'f') => self.push(p | (c - b'a' + 10)),
+            (Escape::Hex(p), b'A'..=b'F') => self.push(p | (c - b'A' + 10)),
+
+            (Escape::Some, b'0'..=b'9') => self.state = Escape::Hex((c - b'0') << 4),
+            (Escape::Some, b'a'..=b'f') => self.state = Escape::Hex((c - b'a' + 10) << 4),
+            (Escape::Some, b'A'..=b'F') => self.state = Escape::Hex((c - b'A' + 10) << 4),
+
+            (Escape::Some, b' ' | b'"' | b'#' | b'=' | b'\\') => self.push(c),
+            (Escape::Some, b'+' | b',' | b';' | b'<' | b'>') => self.push(c),
+
+            (Escape::None, b'\\') => self.state = Escape::Some,
+            (Escape::None, ..) => self.push(c),
+
+            _ => return Err(ErrorKind::Failed.into()),
+        }
+
+        Ok(())
+    }
+
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.bytes
+    }
+}
+
+impl<'a> AttributeTypeAndValue<'a> {
+    /// Parses the hex value in the `OID=#HEX` format.
+    fn parse_der(oid: ObjectIdentifier, val: &str) -> Result<Vec<u8>, der::Error> {
+        // Ensure an even number of hex bytes.
+        let mut iter = match val.len() % 2 {
+            0 => [].iter().cloned().chain(val.bytes()),
+            1 => [0u8].iter().cloned().chain(val.bytes()),
+            _ => unreachable!(),
+        };
+
+        // Decode der bytes from hex.
+        let mut bytes = Vec::with_capacity((val.len() + 1) / 2);
+        while let (Some(h), Some(l)) = (iter.next(), iter.next()) {
+            let mut byte = 0u8;
+
+            for (half, shift) in [(h, 4), (l, 0)] {
+                match half {
+                    b'0'..=b'9' => byte |= (half - b'0') << shift,
+                    b'a'..=b'f' => byte |= (half - b'a' + 10) << shift,
+                    b'A'..=b'F' => byte |= (half - b'A' + 10) << shift,
+                    _ => return Err(ErrorKind::Failed.into()),
+                }
+            }
+
+            bytes.push(byte);
+        }
+
+        // Serialize.
+        let value = Any::from_der(&bytes)?;
+        let atv = AttributeTypeAndValue { oid, value };
+        atv.to_vec()
+    }
+
+    /// Parses the string value in the `NAME=STRING` format.
+    fn parse_str(oid: ObjectIdentifier, val: &str) -> Result<Vec<u8>, der::Error> {
+        // Undo escaping.
+        let mut parser = Parser::new();
+        for c in val.bytes() {
+            parser.add(c)?;
+        }
+
+        // Serialize.
+        let value = Any::new(Tag::Utf8String, parser.as_bytes())?;
+        let atv = AttributeTypeAndValue { oid, value };
+        atv.to_vec()
+    }
+
+    /// Converts an AttributeTypeAndValue string into an encoded AttributeTypeAndValue
+    ///
+    /// This function follows the rules in [RFC 4514].
+    ///
+    /// [RFC 4514]: https://datatracker.ietf.org/doc/html/rfc4514
+    pub fn parse(s: &str) -> Result<Vec<u8>, der::Error> {
```

Comment:
I'm a little confused by the `fn parse` name along with `parse_str` and `parse_der`. These seem to be combination parser/serializer methods?

Reply:
Yes, I agree it is a bit awkward. Basically, RFC 4514 defines a human readable encoding for certificate name types. That is, we want to convert from the name type into a string (this part is easy) and from the string back into the relevant name type. The latter operation is difficult because the result needs to be owned, but the name types borrow references.

The compromise is that we parse the string into an owned buffer. Then we create the named type referencing that owned buffer. Then, because of the lifetime issues, we serialize the named type to der.

In practice it looks like this to avoid the lifetime issues:

```rust
            let der = RdnSequence::parse("CN=foo").unwrap();
            let rdns = RdnSequence::from_der(&der).unwrap();
```

If you have alternative suggestions, I'm all ears.

Reply:
Since it's producing encoded DER, I'd suggest something like `RdnSequence::encode_*`, like `RdnSequence::encode_from_string`

Reply:
I renamed all the public functions to `encode_from_string()`. The internal functions were renamed to `encode_str()` and `encode_hex()`.